File: backend/src/schemas/trainSchedule.ts
import { trainSchema } from './train.js';
import { trainRouteSchema } from './trainRoute.js';
import { stationSchema } from './station.js';

export const scheduleStationSchema = {
  type: 'object',
  properties: {
    id: { type: 'number' },
    trainScheduleId: { type: 'number' },
    stationId: { type: 'number' },
    station: stationSchema,
    arrivalTime: { type: 'string', nullable: true },
    departureTime: { type: 'string', nullable: true },
    sequence: { type: 'number' },
  },
};

export const trainScheduleSchema = {
  type: 'object',
  properties: {
    id: { type: 'number' },
    trainId: { type: 'number' },
    trainRouteId: { type: 'number' },
    train: trainSchema,
    trainRoute: trainRouteSchema,
    date: { type: 'string', format: 'date-time' },
    time: { type: 'string' },
    stationTimes: {
      type: 'array',
      items: scheduleStationSchema,
    },
  },
};

export const trainScheduleWithTimestampsSchema = {
  type: 'object',
  properties: {
    id: { type: 'number' },
    trainId: { type: 'number' },
    trainRouteId: { type: 'number' },
    train: trainSchema,
    trainRoute: trainRouteSchema,
    date: { type: 'string', format: 'date-time' },
    time: { type: 'string' },
    stationTimes: {
      type: 'array',
      items: scheduleStationSchema,
    },
    createdAt: { type: 'string' },
    updatedAt: { type: 'string' },
  },
};

export const createTrainScheduleBodySchema = {
  type: 'object',
  required: ['trainId', 'date', 'time', 'stationTimes'],
  properties: {
    trainId: { type: 'number' },
    date: { type: 'string', format: 'date' },
    time: { type: 'string', pattern: '^([01]?[0-9]|2[0-3]):[0-5][0-9]$' },
    stationTimes: {
      type: 'array',
      minItems: 2,
      items: {
        type: 'object',
        required: ['stationId', 'arrivalTime', 'departureTime'],
        properties: {
          stationId: { type: 'number' },
          arrivalTime: { type: 'string', pattern: '^([01]?[0-9]|2[0-3]):[0-5][0-9]$' },
          departureTime: { type: 'string', pattern: '^([01]?[0-9]|2[0-3]):[0-5][0-9]$' },
        },
      },
    },
  },
};

File: android/app/src/main/java/com/mojahid2021/railnet/home/model/TrainSchedule.java
package com.mojahid2021.railnet.home.model;

import com.google.gson.annotations.SerializedName;

import java.util.List;

public class TrainSchedule {
    public int id;
    public int trainId;
    public int trainRouteId;
    public Train train;
    public TrainRoute trainRoute;
    public String date; // ISO date-time string
    public String time;
    public List<StationTime> stationTimes;
    public String createdAt;
    public String updatedAt;

    public static class Train {
        public int id;
        public String name;
        public String number;
        public int trainRouteId;
        public TrainRoute trainRoute;
        public List<CompartmentAssignment> compartments;
    }

    public static class TrainRoute {
        public int id;
        public String name;
        public int startStationId;
        public int endStationId;
        public SimpleStation startStation;
        public SimpleStation endStation;
        public List<RouteStation> routeStations;
        public String createdAt;
        public String updatedAt;
    }

    public static class RouteStation {
        public int id;
        public int previousStationId;
        public int currentStationId;
        public int nextStationId;
        public double distance;
        public double distanceFromStart;
    }

    public static class CompartmentAssignment {
        public int id;
        public int trainId;
        public int compartmentId;
        public int quantity;
        public Compartment compartment;
        public String createdAt;
        public String updatedAt;
    }

    public static class Compartment {
        public int id;
        public String name;
        @SerializedName("class")
        public String clazz; // 'class' is reserved, map to clazz
        public String type;
        public double price;
        public int totalSeats;
    }

    public static class StationTime {
        public int id;
        public int trainScheduleId;
        public int stationId;
        public SimpleStation station;
        public String arrivalTime; // string time
        public String departureTime;
        public int sequence;
    }

    public static class SimpleStation {
        public int id;
        public String name;
        public String city;
        public double latitude;
        public double longitude;
    }
}
